### ISSUE TITLE
Create self-links in each workspace package's node_modules when not hoisted

### ISSUE THREAD

**Issue by jakebailey** [pnpm/pnpm#6316]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

On DefinitelyTyped, we need to write tests for each `@types` package. In order for that to work, however, the `@types` package needs to be in a `node_modules` for TypeScript to pick it up.

**Describe the solution you'd like**

Provide an option that links every workspace package into its own `node_modules`.

This sort of thing sorta/kinda already happens if you're using yarn or npm workspaces by nature of the fact that they hoist their packages all the way to the root (which puts _every_ workspace package into every other workspace packages' `node_modules`).

**Describe the drawbacks of your solution**

Some tooling may be confused by this recursive loop, however, given npm/yarn make a similar (though larger) loop, I think that tooling may actually be fine.

Additionally, this may not work out for an edge case like:

```json
{
	"name": "foo",
	"version": "2.0.0",
	"dependencies": {
		"foo": "1.0.0"
	}
}
```

Not quite sure what to do about that; I can't use npm or yarn as an example as I don't think either allow packages with duplicate names.

**Describe alternatives you've considered**

The workaround the moment is either to always hoist, or to manually instruct pnpm to do this by setting something like:

```json
{
	"devDependencies": {
		"@types/node": "workspace:."
	}
}
```

Which isn't performant in DT due to #6277 (but I'm working on a fix for it regardless). A version range could be used instead, but that duplicates the version number inside of the package.json for cases where there are multiple DT packages with the same name.

**Comment by SoonIter** [comment]
Maybe you can solve the problem through this? 
```json
{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "foo": "link:."
  }
}
```
or
```json
{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "foo": "workspace:*"
  }
}
```

to hoist like npm or yarn
```txt
# .npmrc
public-hoist-pattern[]=*types*
```

**Comment by jakebailey** [comment]
Yes, that is the workaround I said above in "alternatives", and that's what I'm currently using.

Hoisting all types is generally a bad idea for our use case so I won't be doing that. That and the `[]` syntax totally breaks `npm` and `yarn` (so we can't even get them to say "use pnpm") so I'm avoiding that entirely.

**Comment by jakebailey** [comment]
Though, I had not yet tried the `link:` syntax; I'll try that one out as it probably has better performance until #6317.

**Comment by jakebailey** [comment]
That syntax does appear to work as well; I'll have to switch my current WIP to use that instead of `workspace:.`. Probably better in the long term anyhow.

**Comment by jakebailey** [comment]
Pretty good, though this does expose some performance downsides:

![image](https://user-images.githubusercontent.com/5341706/228647464-eacd08ef-e99e-4499-9dc4-067647bcfa9a.png)

![image](https://user-images.githubusercontent.com/5341706/228647491-8f6916a2-eaa0-4aa9-aba3-3df6a62e1784.png)

Roughly 8 seconds or so extra from these calls.

**Comment by jakebailey** [comment]
The above can be simplified a little to get back a few seconds, at least; `path.relative(path1, path2) === ""` is slower than just resolving them.

**Comment by zkochan** [comment]
Yarn has a similar setting: https://yarnpkg.com/configuration/yarnrc/#nmSelfReferences

**Comment by jakebailey** [comment]
The DefinitelyTyped pnpm transition is happening hopefully very soon, as it unblocks a load of important changes. We're still using devDeps with `"@types/absolute-url": "workspace:."` for now as a workaround but it'd be very cool to get this natively. I suspect that a real implementation would be a lot faster as it'd remove a load of edges from the dependency graph if implemented directly.
